analysis/norms.lean

```diff
@@ -0,0 +1,292 @@
+
+class normed_group (α : Type*) extends add_comm_group α, metric_space α :=
+(norm : α → ℝ)
+(dist_eq : ∀ x y, dist x y = norm (x - y))
+
+
+def norm {G : Type*} [normed_group G] : G → ℝ := normed_group.norm 
+notation `∥` e `∥` := norm e
+
+section normed_group
+variables {G : Type*} [normed_group G] {H : Type*} [normed_group H]
+
+
+lemma norm_dist' { g h : G} : dist g h = ∥g - h∥ :=
+normed_group.dist_eq _ _
+
+@[simp]
+lemma norm_dist { g : G} : dist g 0 = ∥g∥ :=
+by { rw[norm_dist'], simp }
+
+lemma norm_triangle (g h : G) : ∥g + h∥ ≤ ∥g∥ + ∥h∥ :=
+calc ∥g + h∥ = ∥g - (-h)∥             : by simp
+         ... = dist g (-h)            : by simp[norm_dist']
+         ... ≤ dist g 0 + dist 0 (-h) : by apply dist_triangle
+         ... = ∥g∥ + ∥h∥               : by simp[norm_dist']
+
+@[simp]
+lemma norm_nonneg {g : G} : 0 ≤ ∥g∥ :=
+by { rw[←norm_dist], exact dist_nonneg }
+
+lemma norm_zero_iff_zero {g : G} : ∥g∥ = 0 ↔ g = 0 :=
+by { rw[←norm_dist], exact dist_eq_zero }
+
+@[simp]
+lemma zero_norm_zero : ∥(0:G)∥ = 0 :=
+norm_zero_iff_zero.2 (by simp)
+
+lemma norm_pos_iff {g : G} : ∥ g ∥  > 0 ↔ g ≠ 0 :=
+begin
+  split ; intro h ; rw[←norm_dist] at *,
+  { exact dist_pos.1 h },
+  { exact dist_pos.2 h }
+end
+
+lemma norm_le_zero_iff {g : G} : ∥g∥ ≤ 0 ↔ g = 0 :=
+by { rw[←norm_dist], exact dist_le_zero }
+
+
+@[simp]
+lemma norm_neg {g : G} : ∥-g∥ = ∥g∥ :=
+calc ∥-g∥ = ∥0 - g∥ : by simp
+      ... = dist 0 g : norm_dist'.symm
+      ... = dist g 0 : dist_comm _ _
+      ... = ∥g - 0∥ : norm_dist'
+      ... = ∥g∥ : by simp
+
+lemma norm_triangle' (g h : G) : abs(∥g∥ - ∥h∥) ≤ ∥g - h∥ := 
+begin
+  have ng := calc ∥g∥ = ∥g - h + h∥ : by simp
+  ... ≤ ∥g-h∥ + ∥h∥ : norm_triangle _ _,
+  replace ng := sub_right_le_of_le_add ng,
+  have nh := calc ∥h∥ = ∥h - g + g∥ : by simp
+  ... ≤ ∥h - g∥ + ∥g∥ : norm_triangle _ _
+  ... = ∥-(g - h)∥ + ∥g∥ : by simp
+  ... = ∥g - h∥ + ∥g∥ : by { rw [show ∥-(g - h)∥ = ∥g-h∥, from norm_neg] },
+  replace nh := sub_right_le_of_le_add nh,
+  replace nh : -(∥g∥ - ∥h∥) ≤ ∥g - h∥ := by simpa using nh,
+  replace nh : -∥g - h∥ ≤ ∥g∥ - ∥h∥ := by simpa using neg_le_neg nh,
+  exact abs_le.2 ⟨nh, ng⟩
+end
+
+instance prod.normed_group {F : Type*} [normed_group F] : normed_group (G × F) :=
+{ norm := λ x, max ∥x.1∥ ∥x.2∥,
+  dist_eq := assume x y, by { simp, repeat {rw norm_dist'}, simp } }
+  
+
+lemma norm_proj1_le (x : G × H) : ∥x.1∥ ≤ ∥x∥ :=
+begin  have : ∥x∥ = max  (∥x.fst∥) ( ∥x.snd∥) := rfl, rw this, simp[le_max_left], end
+
+lemma norm_proj2_le (x : G × H) : ∥x.2∥ ≤ ∥x∥ :=
+begin  have : ∥x∥ = max  (∥x.fst∥) ( ∥x.snd∥) := rfl, rw this, simp[le_max_right], end
+
+instance fintype.normed_group {ι : Type*} {α : ι → Type*} [fintype ι] [∀i, normed_group (α i)] :
+  normed_group (Πb, α b) :=
+{ norm := λf, finset.maxi finset.univ (λ b, ∥f b∥),
+  dist_eq := assume x y, by finish [norm_dist'] }
+
+lemma tendsto_iff_norm_tendsto_zero {f : G → H} {a : G} {b : H} : (f →_{a} b) ↔ ((λ e, ∥ f e - b ∥) →_{a} 0) :=
```

Comment:
You should write it in terms of a general filter, not only the neighborhoods of a point `a`. For instance, you want to know that a sequence `u_n` tends to `b` at infinity if and only if `||u n - b||` tends to `0` at infinity.

Reply:
Thank you for the suggestion! This was written a long time ago, when I was still afraid of filters (I have used a notation to hide them).